22/7/15(금) 2회차 과제 PR

wygbsr opened the pull request.

wygbsr committed:
First commit

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,98 @@
+        }
+    }
+
+    double calculate(double operand1, String operator, double operand2) throws Exception {
+        switch (operator) {
+            case "+":
+                return operand1 + operand2;
+            case "-":
+                return operand1 - operand2;
+            case "*":
+                return operand1 * operand2;
+            case "/":
+                return (int) divide(operand1, operand2);
+            default:
+                throw new IllegalArgumentException();
+        }
+    }
```
Enum 과 BiFunction 을 활용해서 다른 클래스로 분리할 수 있습니다. 
클래스를 분리해서 역할을 나누어 보세요 :)

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,98 @@
+    double output;
+    String trimmedOutput;
+
+    void bootUp(String input) throws Exception {
```
함수에 public, private 에 대한 키워드가 누락되어 있네요. 명시적으로 붙여주는 것이 좋습니다.

wygbsr replied:
처음 작성 시에 Calculator 객체 외부에서 호출되는 bootUp(String)만 default로 선언하고, Calculator 클래스의 필드들과 나머지 메소드는 private으로 선언하였습니다.
그런데 이 경우 테스트 메소드에서 Calculator의 필드 및 메소드에 접근 불가능해서 private도 default 처리하였습니다. 
혹시 필드와 메소드 모두 protected나 public 처리해 주어야하나요?

findstar replied:
필드와 메소드 모두 명시적으로 public, private, protected 를 지정해주는 것이 좋습니다. 
default 이기 때문에 접근에 어려움이 있다면 구조를 변경하는 쪽으로 리팩토링 해보는 것이 좋겠습니다.
테스트 코드에서는 private 메소드를 호출할 수 없기 때문에 public 메소드를 호출하여 내부의 코드를 테스트하게 됩니다.
만약 다른 구조를 생각한다면 생성자를 활용해볼 수도 있습니다.
```
Calculator calculator = new Calculator(input);
calculator.execute();
```

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,98 @@
+    String trimmedOutput;
+
+    void bootUp(String input) throws Exception {
+        checkIfNullOrEmpty(input);
+        setStringArray(input);
+        calculate(stringArray);
+        print();
```
추상화 레벨이 동일하게 구성되어 있어서 읽기가 좋네요 
잘 하셨습니다. 👍

wygbsr replied:
감사합니다. 
슬랙에 소개해 주신 클린 코드 가이드에서 함수가 **한가지 일**만 하도록 권장하고 있는데, 
그렇다면 프로그램을 작성할 때
1) 위와 같이 하나의 head메소드에서 순서대로 다른 메소드(필드 변경, 연산 수행)들을 호출
2) 체인 식으로 클래스 내 메소드가 꼬리물기 호출(1번에서 2번호출, 2번에서 3번호출...)
3) 서브루틴 제외한 모든 메소드를  클래스 외부에서 호출(main등)
중 어떤 것이 바람직한가요?

findstar replied:
경우에 따라서 모두 다릅니다. 
메소드 체이닝을 선호하는 경우도 있으나, 
여기서는 흐름을 그대로 나타내는 방식이 제일 무난하다고 생각합니다. 
지금은 1번과 같은 형태가 좋겠습니다.

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,98 @@
+        return a / b;
+    }
+
+    void print() {
```
print 보다는 printResult 같은 이름은 어떤가요? 결과를 출력한다는 의미가 더 명확해질 것 같습니다.

wygbsr committed:
리팩토링: enum 추가, 생성자 추가, ArithmeticOperation 클래스에 입출력 기능 추가

 -enum, BiFunction, Optional로 Operator 구현
 -Calculator 클래스 생성자 추가
 -입출력 기능 Calculator에서 분리하여 ArithmeticOperation에 추가
 -명시적 접근지정자 지정

 -접근 지정자 변경에 따라 테스트 코드 재작성

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,84 @@
+    private double output;
+    private String trimmedOutput;
+
+    public Calculator() {
+        this("");
+    }
```
인자가 없는 생성자는 테스트용으로만 사용되는 것 같아요.
테스트에서도 굳이 private 변수를 가지고 있을 필요가 없어보이네요.

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,84 @@
+    public Calculator(String input) {
+        this.input = input;
+    }
+    public String execute() throws Exception {
```
메소드와 메소드 사이에는 라인을 하나 비워주세요 :)

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,84 @@
+        return trimmedOutput;
+    }
+
+    private void checkIfNullOrEmpty() throws Exception {
```
이 메소드는 생성자 에서 처리하는게 어떨까요?

findstar commented on src/main/java/org/sscc/study/Calculator.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    private void calculate() throws Exception {
+        checkInvalidOperand();
+        checkInvalidOperator();
```
마찬가지로 계산할 수 있는지 사전에 검증하는 단계는 생성자에서 처리해도 될것 같아요.

wygbsr committed:
리팩토링: Calculator 생성자 수정, 테스트 코드 수정

 -Calculator 클래스 생성자 수정

 -테스트 코드 @BeforeEach 삭제

wygbsr commented:
커밋에 착오가 있었습니다. 새롭게 커밋할 예정입니다.

wygbsr committed:
리팩토링: Calculator 생성자 수정, 테스트 코드 수정

 -Calculator 클래스 생성자 수정. 유효하지 않은 입력 생성자 단계에서 필터

 -테스트 코드 BeforeEach 삭제
 -arithmetic exception 테스트 코드 추가

wygbsr committed:
Merge remote-tracking branch 'origin/assignment2' into assignment2

# Conflicts:
#	src/main/java/org/sscc/study/Calculator.java
#	src/test/java/org/sscc/study/CalculatorTest.java

wygbsr committed:
import문 수정

wygbsr committed:
구조 리팩터링 및 enum용 테스트코드 작성

 -validationCheck() 작성
 -Operator enum의 calculate 메서드 간소화
 -Operator enum의 테스트 코드 작성

findstar reviewed (approved):
피드백 잘 반영해주셨습니다. 
고생하셨어요 👍
